### ISSUE TITLE
Indexing of operators is not inherited

### ISSUE THREAD

**Issue by mehrhardt** [odlgroup/odl#1526]
ODL operators currently do not inherit whether you can index them, e.g. being a `DiagonalOperator` or `BroadcastOperator`. Examples where it fails are listed below. If there is agreement that this feature is desirable, I could give it a try to implement this in ODL.

> import odl
> 
> X = odl.rn(2)
> X2 = X ** 2
> 
> I = odl.IdentityOperator(X)
> A = odl.DiagonalOperator(I, I)
> 
> #%% Problem 1
> A[0]
> B = odl.IdentityOperator(X2)
> B[0]
> 
> #%% Problem 2
> B = A + A
> B[0]
> 
> #%% Problem 3
> B = 2 * A
> B[0]

**Comment by mehrhardt** [comment]
Before making a proper PR, here is what I propose to solve "Problem 1" above.

>     def __getitem__(self, index):
>         """Get sub-operator by index if domain is a ProductSpace.
> 
>         Parameters
>         ----------
>         index : int or tuple of int
>             A pair of integers given as (row, col).
> 
>         Returns
>         -------
>         operator : `IdentityOperator`
>             If index is an integer, return the IdentityOperator the subspace 
>             given by the index.
> 
>         Examples
>         --------
>         >>> X = odl.rn(2) * odl.rn(3)
>         >>> I = odl.IdentityOperator(X)
>         >>> I[0]
>         IdentityOperator(rn(2))
>         """
>         if isinstance(self.domain, ProductSpace):
>             index = int(index)
>             return IdentityOperator(self.domain[index])
>         
>         else:
>             raise TypeError('Indexing only supported if the domain is a '
>                             'ProductSpace')

**Comment by adler-j** [comment]
I would probably propose that productspace operators override `__add__` instead, it feels more natural that they have a "special way of adding" which in this case would again return a productspace operator? I guess it could cause some confusion though.

**Comment by mehrhardt** [comment]
Good idea, @adler-j! However, I can see your point that it may cause confusion if `A + B` is not of `OperatorSum` type. This would not be the case in my proposed solution:
> If A is indexable and B is indexable:
>     (A + B)[i] = A[i] + B[i]
> else:
>     error
@kohr-h, @aringh, @sbanert anyone of you having a strong opinion on this?

**Comment by kohr-h** [comment]
The problems you point out, @mehrhardt, are somewhat different. In general, for composite operators like `OperatorSum`, `OperatorComp` and `ProductSpaceOperator`, I think it's fine to implement indexing as you suggest. The basic pattern would be, "index the constituents and re-combine the sub-operators". That would cover problems 2 and 3 and would be rather straightforward to implement. (However, it might affect quite a number of classes if done thoroughly.)

Problem 1, however, is again one of those cases where two things are mathematically equivalent, but hard to make equivalent in practice. To "solve" this problem would mean to require any operator defined on a product space to have a way to restrict itself to one or several subspaces. This *might* be possible for most of the operators we have, but would add a lot of complexity. I'm not sure whether that would be worth the effort.

**Comment by mehrhardt** [comment]
@kohr-h, that sounds great! The use-case I have (which is about computing partial derivatives of composite functions) would be solved `OperatorSum`, `OperatorComp`, `OperatorLeftScalarMult` and `ProductSpaceOperator` supported indexing.

@adler-j, are you happy about this, too?
